v2/tools/asoctl/internal/importing/importable_resource.go

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package importing
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/pkg/versions"
+)
+
+// ImportableResource is an interface that wraps a Kubernetes resource that can be imported.
+// Different implementations of this interface will be used for different types of resources.
+type ImportableResource interface {
+	// Name is a unique identifier for this resource
+	Name() string
+
+	// Resource returns the actual resource that has been imported.
+	// Only available after the import is complete (nil otherwise).
+	Resource() genruntime.MetaObject
+
+	// Import does the actual import, updating the Spec on the wrapped resource.
+	// ctx allows for cancellation of the import.
+	// If there are any additional resources that also need to be imported, they should be returned.
+	Import(ctx context.Context) ([]ImportableResource, error)
+}
+
+// importableResource is a core of common data and support methods for implementing ImportableResource
+type importableResource struct {
+	scheme *runtime.Scheme
+}
+
+// createBlankObjectFromGVK is a helper function to create a blank object of from a given GVK.
+func (i *importableResource) createBlankObjectFromGVK(gvk schema.GroupVersionKind) (runtime.Object, error) {
+	obj, err := i.scheme.New(gvk)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create blank resource")
+	}
+
+	obj.GetObjectKind().SetGroupVersionKind(gvk)
+	return obj, nil
+}
+
+// selectVersionFromGK is a helper function to select the latest version of a given GroupKind.
+// The latest stable version will be selected if it exists, otherwise the latest preview version will be selected.
+func (i *importableResource) selectVersionFromGK(gk schema.GroupKind) (schema.GroupVersionKind, error) {
+	knownVersions := i.scheme.VersionsForGroupKind(gk)
+	if len(knownVersions) == 0 {
+		return schema.GroupVersionKind{},
+			errors.Errorf(
+				"no known versions for Group %s, Kind %s",
+				gk.Group,
+				gk.Kind)
+	}
+
+	return i.selectLatestVersion(gk, knownVersions), nil
+}
+
+// selectLatestVersion is a helper function to select the latest from a slice of GroupVersions
+// The latest stable version will be selected if it exists, otherwise the latest preview version will be selected.
+func (i *importableResource) selectLatestVersion(
+	gk schema.GroupKind,
+	knownVersions []schema.GroupVersion,
+) schema.GroupVersionKind {
+	// Sort the versions the same way we do in the generator, for consistency.
```

Comment:
minor: Should we leave a link to the code here (and there?)

Or put the sort order into a function and share that?

Reply:
The sort order is already in a shared function. That's what `versions.Compare()` is.